src/main/java/core/AbstractGameState.java

```diff
@@ -223,6 +215,26 @@ public final double getScore(int playerId) {
         return _getUnknownComponentsIds(playerId);
     }
 
+    /**
+     * Used by ForwardModel.next() to log history (very useful for debugging)
+     *
+     * @param action The action that has just been applied (or is about to be applied) to the game state
+     */
+    protected void recordAction(AbstractAction action) {
+        history.add(action);
+        historyText.add("Player " + this.getCurrentPlayer() + " : " + action.getString(this));
```

Comment:
I wonder if it's worth storing just the action and player ID in the history? And then compute text on the fly as needed, to avoid making the game state too heavy. Although I don't know how much  this influences, and we'd need the game states for the history to print out the actions too... Mm thoughts?

Reply:
Yup - I thought about this too.  The implementation *should* be highly efficient - the key point is that each String is only added once and never, ever copied. The two crucial code optimisations are:
1) in AbstractGameState.copy(), where we explicitly do not copy actions (just their references), or their string representations on the basis that these are now immutable history. This means that the same String is used in all copied future states.
        s.history = new ArrayList<>(history);
        s.historyText = new ArrayList<>(historyText);
            // we do not copy individual actions in history, as these are now dead and should not change

2) In AbstractGameState.equals():
        // we deliberately exclude history from this equality check

So - I think that the memory and processing footprint of this is actually quite slim. (If one deliberately copied all those Strings and Actions each time though....)